### PR TITLE
Implement new vector types

### PR DESCRIPTION
**Pull request by gita-omr**
- see https://github.com/eclipse/omr/pull/6353 for the background
- to enable VectorAPI support, VectorAPIExpansion::supportedOnPlatform needs to be modified
  correspondingly
- to enable specific opcode getSupportsOpCodeForAutoSIMD needs to be modified correspondingly

depends https://github.com/eclipse/omr/pull/6353

### CODE REVIEW

**Comment by gita-omr** [comment]
Fixed copyright.

**Comment by gita-omr** [comment]
@0xdaryl @jdmpapin  @mstoodle @vijaysun-omr could you please review the common part?
@knn-k @Akira1Saitoh @r30shah @BradleyWood @IBMJimmyk could you please review aarch64, z, x, and p files correspondingly?

File: runtime/compiler/il/J9DataTypes.cpp
Comment (mstoodle) [inline runtime/compiler/il/J9DataTypes.cpp:219]: I see this pattern repeated everywhere: would it be better to introduce a query like `isOMRDataType(TR::DataType dt)` and change all these to use that query rather than spraying `|| dt.isVector()` everywhere?

File: runtime/compiler/ilgen/Walker.cpp
Comment (mstoodle) [inline runtime/compiler/ilgen/Walker.cpp:2261]: can `getVectorElementType()` just return `this` if not a vector type?

Comment (gita-omr) [reply runtime/compiler/ilgen/Walker.cpp:2261]: Interesting idea. I worry that there will be some hidden bugs due to that. I think currently, whenever we ask `getVectorElementType() `we indeed expect the type to be a vector. And actually, this method was not added now. It existed before. It used to return `TR::NoType` on non-vector types and now it asserts.

I would like to preserve the current (or more strict)behavior for now as much as possible. But we can definitely consider this change in the future.

File: runtime/compiler/il/J9DataTypes.cpp
Comment (gita-omr) [reply runtime/compiler/il/J9DataTypes.cpp:219]: Great idea. I will add  that query. Thanks!

**Comment by gita-omr** [comment]
Added isOMRDataType().

**Comment by gita-omr** [comment]
Put WIP back since I still need to run Step 1 of the OpenJ9 Committer Guide for coordinated merges.

**Comment by gita-omr** [comment]
Reviewed the code once again on the subject of any tables that are indexed by TR::DataTypes and therefore might need to be changed. Applied some minor fixes.

**Comment by gita-omr** [comment]
Fixed white space.

**Comment by gita-omr** [comment]
The PR is ready for a final review.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11 depends eclipse/omr#6353

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11 depends https://github.com/eclipse/omr/pull/6353

Re-starting because upstream changes were made and some builds stalled in previous attempt.

**Comment by gita-omr** [comment]
> Jenkins test sanity all jdk8,jdk11 depends [eclipse/omr#6353](https://github.com/eclipse/omr/pull/6353)
> 
> Re-starting because upstream changes were made and some builds stalled in previous attempt.

Sorry, I indeed pushed something into the OMR PR.

**Comment by gita-omr** [comment]
All current failures look like this:

`Pipeline] End of Pipeline
java.lang.StringIndexOutOfBoundsException: String index out of range: 7
	at java.base/java.lang.String.substring(String.java:2691)
	at jdk.internal.reflect.GeneratedMethodAccessor263.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.codehaus.groovy.reflection.CachedMethod.invoke(CachedMethod.java:93)
	at groovy.lang.MetaMethod.doMethodInvoke(MetaMethod.java:325)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1213)
	at groovy.lang.MetaClassImpl.invokeMethod(MetaClassImpl.java:1022)
	at org.codehaus.groovy.runtime.callsite.PojoMetaClassSite.call(PojoMetaClassSite.java:47)
	at org.codehaus.groovy.runtime.callsite.CallSiteArray.defaultCall(CallSiteArray.java:48)
	at org.codehaus.groovy.runtime.callsite.AbstractCallSite.call(AbstractCallSite.java:113)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:163)
	at org.kohsuke.groovy.sandbox.GroovyInterceptor.onMethodCall(GroovyInterceptor.java:23)
	at org.jenkinsci.plugins.scriptsecurity.sandbox.groovy.SandboxInterceptor.onMethodCall(SandboxInterceptor.java:158)
	at org.kohsuke.groovy.sandbox.impl.Checker$1.call(Checker.java:161)
	at org.kohsuke.groovy.sandbox.impl.Checker.checkedCall(Checker.java:165)
	at com.cloudbees.groovy.cps.sandbox.SandboxInvoker.methodCall(SandboxInvoker.java:17)
	at Script3.checkout_pullrequest(Script3.groovy:158)
	at Script3.get_sources(Script3.groovy:105)
	at Script3.get_source(Script3.groovy:47)
	at ___cps.transform___(Native Method)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:86)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:113)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:83)
	at jdk.internal.reflect.GeneratedMethodAccessor107.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ContinuationGroup.methodCall(ContinuationGroup.java:89)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.dispatchOrArg(FunctionCallBlock.java:113)
	at com.cloudbees.groovy.cps.impl.FunctionCallBlock$ContinuationImpl.fixArg(FunctionCallBlock.java:83)
	at jdk.internal.reflect.GeneratedMethodAccessor107.invoke(Unknown Source)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.cloudbees.groovy.cps.impl.ContinuationPtr$ContinuationImpl.receive(ContinuationPtr.java:72)
	at com.cloudbees.groovy.cps.impl.ConstantBlock.eval(ConstantBlock.java:21)
	at com.cloudbees.groovy.cps.Next.step(Next.java:83)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:174)
	at com.cloudbees.groovy.cps.Continuable$1.call(Continuable.java:163)
	at org.codehaus.groovy.runtime.GroovyCategorySupport$ThreadCategoryInfo.use(GroovyCategorySupport.java:129)
	at org.codehaus.groovy.runtime.GroovyCategorySupport.use(GroovyCategorySupport.java:268)
	at com.cloudbees.groovy.cps.Continuable.run0(Continuable.java:163)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.access$001(SandboxContinuable.java:18)
	at org.jenkinsci.plugins.workflow.cps.SandboxContinuable.run0(SandboxContinuable.java:51)
	at org.jenkinsci.plugins.workflow.cps.CpsThread.runNextChunk(CpsThread.java:185)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.run(CpsThreadGroup.java:402)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup.access$400(CpsThreadGroup.java:96)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:314)
	at org.jenkinsci.plugins.workflow.cps.CpsThreadGroup$2.call(CpsThreadGroup.java:278)
	at org.jenkinsci.plugins.workflow.cps.CpsVmExecutorService$2.call(CpsVmExecutorService.java:67)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at hudson.remoting.SingleLaneExecutorService$1.run(SingleLaneExecutorService.java:139)
	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
	at jenkins.security.ImpersonatingExecutorService$1.run(ImpersonatingExecutorService.java:68)
	at java.base/java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:515)
	at java.base/java.util.concurrent.FutureTask.run(FutureTask.java:264)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:836)`

**Comment by gita-omr** [comment]
Jenkins test sanity all jdk8,jdk11 depends https://github.com/eclipse/omr/pull/6353

**Comment by gita-omr** [comment]
Jenkins test sanity all jdk8,jdk11 depends eclipse/omr/#6353

**Comment by knn-k** [comment]
Builds failed with `ERROR: Bad Depends Repo: 'omr/' in 'eclipse/omr/#6353'`.

**Comment by knn-k** [comment]
Jenkins test sanity all jdk8,jdk11 depends eclipse/omr#6353

**Comment by gita-omr** [comment]
Thanks @knn-k !

**Comment by 0xdaryl** [comment]
Reviewers:  Last call for review.

@mstoodle: could you formally approve this PR if you are satisfied with @gita-omr's resolution of your comments?

**Comment by 0xdaryl** [comment]
Lest the information be lost, this PR passed the acceptance test criteria for JDK8 and JDK11.

File: runtime/compiler/ilgen/Walker.cpp
Comment (mstoodle) [reply runtime/compiler/ilgen/Walker.cpp:2261]: All right. It just seemed like a common pattern that could be simplified, but you should probably change the name of it if you're going to follow my suggestion anyway (since it's a little weird to ask a non-vector type for its vector element type, unless you consider all types to be vectors and scalars are just a degenerate vector. but i don't think that's how you consistently view them so...). Anyway, I can live with a follow-on change.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11 depends https://github.com/eclipse/omr/pull/6353

Re-running because the previous acceptance run occurred over a week ago.

**Comment by 0xdaryl** [comment]
Jenkins test sanity all jdk8,jdk11 depends https://github.com/eclipse/omr/pull/6353

**Comment by knn-k** [comment]
Jenkins test sanity all jdk8,jdk11 depends eclipse/omr#6353

**Comment by knn-k** [comment]
Build failure in https://openj9-jenkins.osuosl.org/job/Build_JDK11_x86-64_linux_Personal/348:
```
[2022-03-30T04:56:12.595Z] /home/jenkins/workspace/Build_JDK11_x86-64_linux_Personal/omr/compiler/x/codegen/FPTreeEvaluator.cpp: In static member function 'static TR::Register* OMR::X86::TreeEvaluator::vdsqrtEvaluator(TR::Node*, TR::CodeGenerator*)':
[2022-03-30T04:56:12.595Z] /home/jenkins/workspace/Build_JDK11_x86-64_linux_Personal/omr/compiler/x/codegen/FPTreeEvaluator.cpp:575:47: error: 'VectorDouble' is not a member of 'TR'
[2022-03-30T04:56:12.595Z]     TR_ASSERT_FATAL(node->getDataType() == TR::VectorDouble, "Unsupported datatype for vdsqrt opcode");
```

**Comment by knn-k** [comment]
Jenkins test sanity all jdk8,jdk11 depends eclipse/omr#6353

**Comment by knn-k** [comment]
Jenkins test sanity s390x_linux jdk8 depends eclipse/omr#6353

The previous job failed with `ub18-390-1 was marked offline: Connection was broken: java.io.EOFException` in https://openj9-jenkins.osuosl.org/job/Test_openjdk8_j9_sanity.functional_s390x_linux_Personal_testList_0/131/

**Comment by 0xdaryl** [comment]
Z failure is infrastructural.  All testing was successful.

**Comment by 0xdaryl** [comment]
Jenkins test sanity plinux,aix jdk8,jdk11 depends eclipse/omr#6353

Previous acceptance testing was successful.  An upstream OMR change specific to Power was merged, so relaunching acceptance on just those affected platforms.

**Review by 0xdaryl — APPROVED** [review]

**Pull request merged**
